services/blockstorage/service.go

```diff
@@ -171,14 +187,49 @@ func (s *service) GetResultsBlockHeader(input *services.GetResultsBlockHeaderInp
 	return s.loadResultsBlockHeader(input.BlockHeight)
 }
 
+func (s *service) createEmptyTransactionReceiptResult() *services.GetTransactionReceiptOutput {
+	return &services.GetTransactionReceiptOutput{
+		TransactionReceipt: nil,
+		BlockHeight:        s.lastCommittedBlockHeight(),
+		BlockTimestamp:     s.lastCommittedBlockTimestamp(),
+	}
+}
+
 func (s *service) GetTransactionReceipt(input *services.GetTransactionReceiptInput) (*services.GetTransactionReceiptOutput, error) {
-	panic("Not implemented")
+	start := input.TransactionTimestamp - primitives.TimestampNano(s.config.BlockTransactionReceiptQueryStartGraceSec().Nanoseconds())
+	end := input.TransactionTimestamp + primitives.TimestampNano((s.config.BlockTransactionReceiptQueryEndGraceSec() + s.config.BlockTransactionReceiptQueryTransactionExpireSec()).Nanoseconds())
+
+	blocksToSearch := s.persistence.GetBlocksByTimeRange(start, end)
```

Comment:
The function `GetBlocksByTimeRange` is very expensive in memory. Blocks will be very big, around 1MB each.. If the time range we search in is 30 minutes (transaction expiry), and we close a block every 5 seconds, then this function returns over 300MB of memory.. is my calculation correct?

I'm not sure we can read from our DB to memory 300MB for every query.. it will make the node vulnerable to DoS

We can optimize this significantly.. for example, we only need the `ResultsBlock.Headers` in step 1, which are quite small. And only after we find the positively matched blocks by bloom, we can query the transactions from the persistent storage (the transactions are the big memory hit - they are big).

If we decide it's too early to optimize at this stage, at least add a very clear TODO that this code cannot go to production before being changed.

Reply:
todo was added a few minutes ago, its all true and we know this is problematic, lets wait for real persistence and then protect ourselves accordingly - this was refactored already to make it more safe regardless of the persistence